xdem/coreg.py

```diff
@@ -892,13 +892,15 @@ class Deramp(Coreg):
     Estimates an n-D polynomial between the difference of two DEMs.
     """
 
-    def __init__(self, degree: int = 1):
+    def __init__(self, degree: int = 1, max_points: int = 5e5):
```

Comment:
Nice! It kind of reinvents the `subsample` argument for `self.fit()`. On the other hand, the `subsample` argument cannot currently subsample differently for different parts in the pipeline. 

I'll add an issue for that, and we can have it like you implemented right now (but it would be nice to have automatically for all `Coreg` subclasses!)

Reply:
+1

Reply:
My idea was that this argument would be used by the subsampling functionality directly. I could just rename "max_point" into "subsample" if that makes it more consistent.

Reply:
Sounds good!

Reply:
I added the `Coreg` implementation as an issue in #137

Reply:
It would be good to have an optional argument that'd make a call to a subsampling function that is the same everywhere (we have one in `spstats_tools` too).  For instance =None would mean no subsampling.
Maybe it would best fit in `spatial_tools`?

Reply:
That's a great idea!

Reply:
Yes, that would be even better!

Reply:
Damn, that function is harder than I expected to write! 
I want to have an option to return the indices of the array that are randomly extracted (for example if willing to extract the same random sample from different DEMs), but I have issues getting those indexes in a general case for a 1D to 3D array, when there are masked values.
This is what I have so far:
```
def subsample_raster(array: Union[np.ndarray, np.ma.masked_array], subsample: Union[float, int],
                     return_indices: bool = False) -> np.ndarray:
    """
    Randomly subsample a 1D or 2D array by a subsampling factor, taking only non NaN/masked values.

    :param subsample: If < 1, will be considered a fraction of valid pixels to extract.
If > 1 will be considered the number of pixels to extract.
    :param return_indices: If set to True, will also return the extracted indices.
 
    :returns: The subsampled array (1D)
    """
    # Get number of points to extract
    if (subsample <= 1) & (subsample >= 0):
        npoints = subsample * np.size(array)
    elif subsample > 1:
        npoints = int(subsample)

    # Remove invalid values and flatten array
    orig_shape = array.shape
    mask = get_mask(array)
    array = array[~mask]

    # Checks that array and npoints are correct
    assert np.ndim(array) == 1
    if npoints >= len(array):
        npoints = len(array)

    # Randomly extract npoints without replacement
    indices = np.random.choice(len(array), npoints, replace=False)
```
But `indices` is for now a flat array, whose length is not necessarily equal to the number of pixels in `array` or `mask`... I don't know how to easily get back to the original pixels coordinates. The easiest I see is to create an array containing the pixel position in each dimension, but for a 3D array, I would basically create three arrays of the same size as my input, which could possibly use a lot of memory!
Any ideas welcome.

Reply:
The already existing subsample script solves this I think:
https://github.com/GlacioHack/xdem/blob/74cf3528daa3bc35b853281e66dad05eb1b7a2f8/xdem/coreg.py#L498

It takes a random choice of row/col indices that are finite.

Reply:
I didn't know this existed! I need to check that out. Anyway, this should be a separate function as we discussed.

The issue with how it is implemented now is that it masks pixels. This raises two issues:
- it does not reduce the number of points, so we're possibly duplicating the data. Ideally, we should pass a flattened 1D array to the underlying _fit_func, but then they all need to accept 1d arrays. This will not work for Nuth&Kaab yet as we need a raster to calculate the slope/aspect.
- masking random pixels will have a terrible effect on the Nuth & Kaab as the slope/aspect need to be calculated, which involve using neighbor pixels that might possibly be masked...

Reply:
This also explains my question to you regarding the inliers_mask @erikmannerfelt. Currently, by default subsample = 1, which means inlier_mask is never used!

Reply:
There's an issue with your subsampling method... rows and cols are reversed!
This is why we should have a separate function for this and test it.
Also, for now it is only implemented for a 2D case. I would like it to be more generic.

Reply:
I found an easy way to code this subsampling function. I will add it as an issue and work on it in a different PR.